content/testing.md

```diff
@@ -0,0 +1,683 @@
+
+Of note here:
+
+ - Before running, each test sets up the data it needs using the `generateData` helper (see [the section on creating integration test data](#creating-integration-test-data) for more detail) then goes to the homepage.
+
+ - Although Flow Router doesn't take a done callback, we can use `Tracker.afterFlush` to wait for all it's reactive consequences to occur.
+
+ - Here we wrote a little utility (which could be abstracted into a general package) to wait for all the subscriptions which are created by the route change (the `todos.inList` subscription in this case) to become ready before checking their data.
+
+ - The `catchAsync` utility is useful because otherwise we don't see the results of failures inside callbacks. XXX: perhaps this test should be rewritten with promises to avoid this?
+
+
+<h3 id="running-full-app-tests">Running full-app tests</h3>
+
+To run the [full-app tests](#test-modes) in our application, we run:
+
+```
+meteor test --full-app --driver-package avital:mocha
+```
+
+When we connect to the test instance in a browser, we want to render a testing UI rather than our app UI, so the `mocha-web-reporter` package will hide any UI of our application and overlay it with its own. However the app continues to behave as normal, so we are able to route around and check the correct data is loaded.
+
+<h3 id="creating-integration-test-data">Creating data</h3>
+
+To create test data in full-app test mode, it usually makes sense to create some special test methods which we can call from the client side. Similar to the way we cleared the database using a method in the `beforeEach` in the [test data](#test-data) section above, we can call that method before running our tests. 
+
+In the case of our routing tests, we've used a file called `imports/api/generate-data.app-tests.js` which defines this method (and will only be loaded in full app test mode, so is not available in general!):
+
+```js
+// This file will be auto-imported in the app-test context, ensuring the method is always available
+
+import { Meteor } from 'meteor/meteor';
+import { Factory } from 'meteor/factory';
+import { resetDatabase } from 'meteor/xolvio:cleaner';
+import { Random } from 'meteor/random';
+import { _ } from 'meteor/underscore';
+
+const createList = (userId) => {
+  const list = Factory.create('list', { userId });
+  _.times(3, () => Factory.create('todo', { listId: list._id }));
+  return list;
+};
+
+Meteor.methods({
+  generateFixtures() {
+    resetDatabase();
+
+    // create 3 public lists
+    _.times(3, () => createList());
+
+    // create 3 private lists
+    _.times(3, () => createList(Random.id()));
+  },
+});
+
+let generateData;
+if (Meteor.isClient) {
+  // Create a second connection to the server to use to call test data methods
+  // We do this so there's no contention w/ the currently tested user's connection
+  const testConnection = Meteor.connect(Meteor.absoluteUrl());
+
+  generateData = () => Promise.denodeify((cb) => {
```

Comment:
Yeah this is the downside of using promises -- Meteor code isn't really setup for them -- as we know from `testAsyncMulti` there isn't really a great way to write async code in Meteor..